common/src/market/impl.rs

```diff
@@ -88,48 +101,33 @@ impl Market {
     fn snapshot_with_yield_distribution(&mut self, yield_distribution: BorrowAssetAmount) -> u32 {
         let time_chunk = self.configuration.time_chunk_configuration.now();
 
-        if let Some((last_index, old_snapshot)) =
-            self.snapshots.len().checked_sub(1).and_then(|last_index| {
-                self.snapshots
-                    .get(last_index)
-                    .filter(|s| s.time_chunk == time_chunk)
-                    .map(|s| (last_index, s))
-            })
-        {
-            let new_snapshot = Snapshot {
+        // If still in current time chunk, just update the current snapshot.
+        if self.current_snapshot.time_chunk == time_chunk {
+            self.current_snapshot.end_timestamp_ms = env::block_timestamp_ms().into();
+            self.current_snapshot.deposited = self.borrow_asset_deposited;
+            self.current_snapshot.borrowed = self.borrow_asset_borrowed;
+            self.current_snapshot
+                .yield_distribution
+                .join(yield_distribution);
+        } else {
+            // Otherwise, finalize the current snapshot and create a new one.
+            let mut snapshot = Snapshot {
                 time_chunk,
-                timestamp_ms: old_snapshot.timestamp_ms,
+                yield_distribution,
                 deposited: self.borrow_asset_deposited,
                 borrowed: self.borrow_asset_borrowed,
-                yield_distribution: {
-                    let mut y = old_snapshot.yield_distribution;
-                    y.join(yield_distribution);
-                    y
-                },
+                end_timestamp_ms: env::block_timestamp_ms().into(),
             };
-            self.snapshots.replace_last(new_snapshot);
-            return last_index;
-        }
-
-        let index = self.snapshots.len();
-        let new_snapshot = Snapshot {
-            time_chunk,
-            timestamp_ms: env::block_timestamp_ms().into(),
-            deposited: self.borrow_asset_deposited,
-            borrowed: self.borrow_asset_borrowed,
-            yield_distribution,
-        };
-        self.snapshots.push(new_snapshot);
-        if let Some(previous_snapshot_index) = index.checked_sub(1) {
-            if let Some(previous_snapshot) = self.snapshots.get(previous_snapshot_index) {
-                MarketEvent::SnapshotFinalized {
-                    index: previous_snapshot_index,
-                    snapshot: previous_snapshot.clone(),
-                }
-                .emit();
+            std::mem::swap(&mut snapshot, &mut self.current_snapshot);
```

Comment:
Nice one, I thought about using it but didn't get to refactor the code before you got to it!